aten/src/ATen/core/op_registration/op_registration_test.cpp

```diff
@@ -219,4 +220,265 @@ TEST(OperatorRegistrationTest, givenOpWithMultipleKernels_whenKernelsHaveSameDis
   }, "Tried to register multiple kernels with same dispatch key");
 }
 
+
+/**
+ * This is used to check that a given type works correctly when passed as input
+ * to or as output from a kernel.
+ *
+ * Call ArgTypeTestKernel<Input, Output>::test(input, inputExpectation, output, outputExpectation, schema)
+ * to test that a kernel with `Input` as input type and `Output` as output types,
+ * when called with `input` fulfills `inputExpectation` inside the kernel, then
+ * returns `output` and the returned value fulfills `outputExpectation`.
+ *
+ * Optionally, you can specify the argument list part of a function schema
+ * (e.g. "(Tensor a) -> Tensor") as an additional argument to use when
+ * registering the kernel.
+ */
+template<class InputType, class OutputType = InputType>
+struct ArgTypeTestKernel final : OperatorKernel {
+  explicit ArgTypeTestKernel(InputType input, std::function<void(const InputType&)> inputExpectation, OutputType output)
+  : input_(std::move(input)), inputExpectation_(std::move(inputExpectation)), output_(std::move(output)) {}
+
+  OutputType operator()(InputType input) const {
+    inputExpectation_(std::move(input));
+    return output_;
+  }
+
+  static void test(InputType input, std::function<void(const InputType&)> inputExpectation, OutputType output, std::function<void(const IValue&)> outputExpectation, const std::string& schema = "") {
```

Comment:
Why does `OutputExpectation` take a `IValue` and not an `OutputType`?

Reply:
I guess... because it's inconvenient to cast the IValue back into OutputType? I guess that's fine.

Reply:
error messages would be much worse if say the number of outputs doesn't match because it wouldn't fail in a nice `EXPECT_EQ`, but instead it would fail in some template metaprogramming logic that unpacks the outputs from the stack into a `std::tuple`